In [1]:
import os,sys
import pandas as pd

In [2]:
project_directory=(os.path.dirname(os.getcwd()))
project_directory

'/home/ankit/amazone-phone-machinelearning_analysis'

In [3]:
file_path=os.path.join(project_directory,"Data","processed_data","unzip_data","phone search.csv")

In [4]:
if os.path.exists(file_path):
    df=pd.read_csv(file_path)

else:
    print("Provided path is not exists.")

In [5]:
df.describe()

product_star_rating  product_num_ratings  product_num_offers  \
count           337.000000           340.000000          340.000000   
mean              4.087537          2744.050000            8.426471   
std               0.383544          7331.816833           10.798389   
min               1.600000             0.000000            1.000000   
25%               3.900000           135.250000            1.750000   
50%               4.100000           503.500000            4.000000   
75%               4.300000          1874.250000           11.000000   
max               5.000000         64977.000000           83.000000   

       unit_count  
count    4.000000  
mean     3.500000  
std      1.732051  
min      2.000000  
25%      2.000000  
50%      3.500000  
75%      5.000000  
max      5.000000

In [14]:
df["product_title"].value_counts()

Moto G Play 2023 3-Day Battery Unlocked Made for US 3/32GB 16MP Camera Navy Blue                                                                                                                                      2
OnePlus Nord N30 5G | Unlocked Dual-SIM Android Smart Phone | 6.7&quot; LCD Display | 8 +128GB | 5000 mAh Battery | 50W Fast Charging | 108MP Camera | Chromatic Gray                                                 2
Panasonic Expandable Corded Phone System, Bluetooth Pairing for Wireless Headphones and Hearing Aids, Smart Call Block, KX-TGF892B (Black), 2 Cordless + 1 Corded Handset (Pack of 1)                                 2
Motorola Moto G04 XT2421-2 (Dual Sim) Factory GSM Unlocked | 4+128GB | 6.56&quot; 90Hz LCD Display | 16MP Camera | International Model - (Green)                                                                      2
Nothing Phone (2) - 512 GB + 12 GB Ram, Glyph Interface, OS 2.0, 50 MP Dual Camera,OIS, 6.7” LTPO OLED Display, 4700mAh Battery, Water R

In [7]:
text_input_file_path=os.path.join(project_directory,"Data","processed_data","unzip_data","column_descriptors.txt")

In [8]:
columns_describtion=[]
with open(text_input_file_path,"r") as file:
    columns_describtion=file.readlines()

In [9]:
columns_describtion

['Column Descriptors:\n',
 'asin: Amazon Standard Identification Number (unique product ID).\n',
 'product_title: Title or name of the product.\n',
 'product_price: Current price of the product.\n',
 'product_original_price: Original price of the product (before any discounts).\n',
 'currency: Currency in which the price is listed (e.g., USD).\n',
 'product_star_rating: Star rating given to the product based on customer reviews.\n',
 'product_num_ratings: Number of customer ratings for the product.\n',
 'product_url: URL of the product page on Amazon.\n',
 "product_photo: Link to the product's image.\n",
 'product_num_offers: Number of offers available for the product.\n',
 'product_minimum_offer_price: Minimum price for all offers available.\n',
 'is_best_seller: Boolean flag indicating if the product is a best seller.\n',
 'is_amazon_choice: Boolean flag indicating if the product is Amazon’s Choice.\n',
 'is_prime: Boolean flag indicating Prime eligibility.\n',
 'product_availability

In [10]:
from analysis.analysis_src.Basic_insights_analysis import Basic_Insights_analyze_Strategies,Basic_Insights,Basic_Info_Strategy,Statistical_Info_Strategy

In [11]:
basic_insights=Basic_Insights(strategy=Basic_Info_Strategy())
basic_insights.execute_strategy(df)

shape of the data set is :(340, 22)
************************************************************
5 rows sample data look like this : 

           asin                                      product_title  \
139  B07HB15NMK  AT&amp;T EL52513 5-Handset Cordless Phone with...   
325  B0BJH9DSDV  SIMBROS Unlocked FLIP 4 AT&amp;T SMARTFLIP IV ...   
259  B0CK2RLX1N  Motorola Edge + Plus | 2022 | US Version| 5G S...   
289  B07P936C7F  Yealink SIP-T57W IP Phone - Corded/Cordless - ...   
98   B0CTW8TXGH  Tracfone | Motorola Moto g Play 2024 | Locked ...   

    product_price product_original_price currency  product_star_rating  \
139       $106.08                    NaN      USD                  4.3   
325        $69.99                    NaN      USD                  3.5   
259       $184.95                $219.00      USD                  4.0   
289       $202.00                $359.00      USD                  3.7   
98         $49.99                    NaN      USD                  4.3   


In [12]:
basic_insights.set_strategy(Statistical_Info_Strategy())
basic_insights.execute_strategy(df)

Statistical summary for numerical columns/features
       product_star_rating  product_num_ratings  product_num_offers  \
count           337.000000           340.000000          340.000000   
mean              4.087537          2744.050000            8.426471   
std               0.383544          7331.816833           10.798389   
min               1.600000             0.000000            1.000000   
25%               3.900000           135.250000            1.750000   
50%               4.100000           503.500000            4.000000   
75%               4.300000          1874.250000           11.000000   
max               5.000000         64977.000000           83.000000   

       unit_count  
count    4.000000  
mean     3.500000  
std      1.732051  
min      2.000000  
25%      2.000000  
50%      3.500000  
75%      5.000000  
max      5.000000  
************************************************************
Statistical summary for categorical columns/features 
              

# Basic Insights About Data >
- ## **Shape** -> Having 340 rows and 22 columns.

- ## **Columns Description**:
    - **asin**: Amazon Standard Identification Number (unique product ID).
    - **product_title**: Title or name of the product.
    - **product_price**: Current price of the product.
    - **product_original_price**: Original price of the product (before any discounts).
    - **currency**: Currency in which the price is listed (e.g., USD).
    - **product_star_rating**: Star rating given to the product based on customer reviews.
    - **product_num_ratings**: Number of customer ratings for the product.
    - **product_url**: URL of the product page on Amazon.
    - **product_photo**: Link to the product's image.
    - **product_num_offers**: Number of offers available for the product.
    - **product_minimum_offer_price**: Minimum price for all offers available.
    - **is_best_seller**: Boolean flag indicating if the product is a best seller.
    - **is_amazon_choice**: Boolean flag indicating if the product is Amazon’s Choice.
    - **is_prime**: Boolean flag indicating Prime eligibility.
    - **product_availability**: Availability status of the product.
    - **climate_pledge_friendly**: Indicates if the product is marked as Climate Pledge Friendly.
    - **sales_volume**: Estimated sales volume for the product.
    - **delivery**: Information on delivery options and times.
    - **has_variations**: Boolean flag indicating if the product has variations (e.g., size, color).
    - **unit_price**: Price per unit for products sold in bulk or by volume.
    - **unit_count**: Number of units in the product listing.
    - **coupon_text**: Information on available coupons or discounts.

## Feature/Columns type and Statistical Summary
- # 1. Numerical columns :
    - **product_star_rating**:
        - **Type** -> "float64"
        - **missing values** -> 3 rows have missing values nan
        - **Statistics Ratings summary** :
            - range -> 1 to 5.
            - mean -> 4.1 rating with standard devation 0.38 which is a best rating 
        - **Conclusion** : The data is left skewed data. and having outliers.

    - **product_num_ratings**: Number of customer rating for the product.
        - **Type** -> "int64"
        - **missing values** -> zero (0)
        - **Statistics summary**:
            - **max number** : total number of people doing rating is 64977 over 340 type of product.
            - **mean** : mean is 2744 with 7331 standard deviation.
        - **conclusion** : right skewed data distribution and having outliers on right hand side.

    - **product_num_offers**: Number of offers available for the product.
        - **Type** -> "int64"
        - **missing values** -> zero (0)
        - **Stratistics summary** :
            - **min offer** :1
            - **max offer** : 83 which is a big number and seems as a outliers.
            - **mean** : around 8 offer is for every item and it increase with 10.8 standard devation number.
        - **Conclusion**: Having outliers and distribution is highly right skewed data as aspect of 83 offers.
    
    - **unit_count**: Number of units in the product listing.
        - **Type** -> "float64"
        - **missing values** -> 336 out of 340.
        
        - **Conclusion** : Most values are NaN, So it seems useless, Operation should be performed is drop will be right decision.

- # 2. Categorical columns:
    - **asin** : Amazon Standard Identification Number (a unique product id).
        - **Type** -> "object" which represent a string.
        - **missing values** -> zero (0)
        - **Statistics summary**:
            - **unique values** : 315 which means 25 item uses the same id out of 315 items or may they are miss placed.
            - **top frequency item name and frequency** item name is "B0BQ118F2T" with frequency 2 times.
        - **Conclusion** : mostly frequecy is 2 means 25 items are unique(in ununique asin id.) but they uses the unique id out of unique item in 315 number.

    - **product_title**: title or name of product.
        - **Type** -> "object" which represent a string 
        - **missing values** -> zero(0)
        - **Statistics summary**:
            - **unique title** : 315 have unique title , and another 25 are  using in of those title.
            - **top title name whcih have top frequency** : title name is :Moto G Play 2023 3-Day Battery Unlocked Made for US 3/32GB 16MP Camera Navy Blue which has frequency 2. and have many more or 24 items that has 2 frequency.
        - **Conclusion**: asin and product_tile looks like co-releated, based on concept or scenario like have unique 315 and 25 uses anothers  asin or product_title name.

    - **product_price**: Current price of the product.
        - **Type** -> "object" means a string because it used currency sign like $106.08 (dollar sign)
        - **missing value** -> 4 , a small number if change its types, to replace with $ then we can use mean and medium method or k-means neighbour like techniques.
        - **Statistics summary**:
            - **unique value**-> 241 items have unique items, another uses or have different price, for price this is not a problem.
            - **same price unit and frequency**-> top frequncy is "12" and unit is "$99.99".
        - **Conclusion**: If the curreny is only one unique values then remove or replace the dollar sign with "" and make it float type, so that fill the values in a better way or in an accurate price.

    - **product_original_price** : Product original price before any discounts.
        - **Type** -> "object" , because it also inculde currency sign like dollar example $119.99.
        - **missing values** -> 192 entries are missing (NaN) which is 56% with respect to entire this column data.
        - **Statistics Summary**:
            - **Unique values** -> 108 items have unique original price.
            - **most frequent price and frequency** ->$119.99 price occurs 8 times. not a problem.
        - **Conclusion** : half entires or original price is missing, so first try to convert it to int or float if possible, then try like mean or meduim operation to fill null values.
    
    - **currency** : In which price is listed (like USD).
        - **Type** -> "object" a string may be a sign or in charactor(words).
        - **missing values** -> 4 values are missing.
        - **most frequent** -> There is only frequecy USD (DOLLAR $)
        - **Conclusion**: Helps in price features like original price and so on because they seems very corelated each other.

    - **product_url** : URL for the product page on AMAZON.
        - **Type** -> "obect" means string.
        - **missing values** -> Zero (0)
        - **Unique values** -> 315, 25 have duplicate values.
        - **top frequency** -> 2 because 25 is duplicated data.
        - **Conclusion** : 25 are duplicated data and rest are unique.

    - **product_photo**: Links to the product image.
        - **Type** ->"object" means a string .
        - **missing values** -> Zero (0)
        - **Unique values** 304, rest images link may be uses same,not a problem , becuase one product may be use like phone, mobile with different variants(different configuration).
        - **Conculsion** : all entries are right but not uses in training the model, just uses only for refrene or giving inforatioin(messages) about products.

    - **product_minimum_offer_price** : Minimum price of product for all over offers.
        - **Type** -> "object" means string because must include the currency sign or curreny represent sign.
        - **missing values** -> 4 values are missing.
        - **Unique price** -> 273 items have unique price.
        - **top same frequency and value** -> $99.99 price is most frequent price and the frequency is  7 times.

    - **product_availability** : Availability status of the product.
        - **Type** -> "object" represent the string.
        - **missing_values** -> 265 entries are missing arouund 78 % data is missing,reason define in conclusion.
        - **top frequent value and number** -> Only 1 left in stock - order soon  got 13 times , and most frequent or values are like this only that number left in stock.
        - **Conclusioin** - lot of missing value it may there are many product over demand, and there is also one scenario that is there is no product availability.

    - **sales_volume** :Estimated sales volume for the product.
        - **Type** -> "object" represent the string
        - **missing_values** -> 20 entries have missing value.
        - **top frequent value** -> "100+ bought in past month" string is mostly used like around 49 times.
        - **Conclusion** -> This represent the sales and by seeing this customer can be get attraction.

    - **delivery** :Information on delivery options and charges.
        - **Types** -> "object" represent the string.
        - **missing values** -> 3 entries are NaN, 337 have values.
        - **Unique values** -> 92 are unqiue sting.
        - **most frequent values** -> "FREE delivery Tue, Sep 24" comes more time with respect to others and frequency is 142.

        - **Conclusion** : This repesent the data and a month and year, if i used it for making a recommender system then should remove the year and month in these columns.

    -



- # 3. Bool type columns:
    - **is_best_seller** ->Boolean flag indicating if the product is a best seller.
        - **Type** -> bool (include two values (True or False) or (0 or 1))
        - **missing values** -> Zero (0)
        - **most frequent value** -> False with frequency 338 and rest True get 2 frequency.
    - **is_amazon_choice** ->Boolean flag indicating if the product is Amazon's choice.
        - **Type** -> bool 
        - **missing values** -> Zero(0)
        - **most frequent value(top)** -> False with frequency 338 , and True frequency is 2.

    - **is_prime** -> Boolean flag indicating the  prime eligiblity.
        - **Type** -> bool
        - **missing values** -> Zero(0)
        - **most frequent value(top)** -> Ture got 248 and flase got 92 in data.

In [25]:
df.delivery.value_counts()

FREE delivery Tue, Sep 24                                                                     142
FREE delivery Tue, Sep 24 on $35 of items shipped by Amazon                                    36
FREE delivery Wed, Sep 25                                                                      18
FREE delivery                                                                                  14
FREE delivery Tomorrow, Sep 20                                                                  7
                                                                                             ... 
FREE delivery Thu, Sep 26 on $35 of items shipped by AmazonOr fastest delivery Tue, Sep 24      1
FREE delivery Wed, Oct 2 Only 6 left in stock - order soon.                                     1
FREE delivery Sep 20 - 23 Only 1 left in stock - order soon.                                    1
FREE deliveryOnly 10 left in stock - order soon.                                                1
FREE delivery Sep 25